hasktorch/src/Torch/Random.hs

```diff
@@ -33,53 +34,68 @@ import qualified Torch.Internal.Type as ATen
+  } deriving (Eq, Show)
 
 mkGenerator :: Device -> Word64 -> IO Generator
 mkGenerator device seed =
   case device of
     Device CPU _ -> do
       genPtr <- ATen.newCPUGenerator seed
-      genenerator <- newTVarIO (Just genPtr)
-      return $ Generator device seed genenerator
+      genenerator <- newTVarIO (Right genPtr)
+      return $ UnsafeGenerator genenerator
     Device CUDA idx -> do
       genPtr <- ATen.newCUDAGenerator (fromIntegral idx)
       ATen.generator_set_current_seed genPtr seed
-      genenerator <- newTVarIO (Just genPtr)
-      return $ Generator device seed genenerator
+      genenerator <- newTVarIO (Right genPtr)
+      return $ UnsafeGenerator genenerator
 
 type RandomGenFunc = ForeignPtr ATen.IntArray -> ForeignPtr ATen.Generator -> ForeignPtr ATen.TensorOptions -> IO (ForeignPtr ATen.Tensor)
 
+
 generatorFactory :: RandomGenFunc -> [Int] -> TensorOptions -> Generator -> (Tensor, Generator)
-generatorFactory func size options Generator {..} =
+generatorFactory func size options (UnsafeGenerator generator) =
   unsafePerformIO $ do
     mGenerator <- atomically $ do
       v <- readTVar generator
       case v of
-        Just v' -> do
-          writeTVar generator Nothing
-          return $ Just v'
-        Nothing -> return Nothing
+        Right v' -> do
+          let device =
+                if generatorIsCuda v' then
+                  Device {deviceType = CUDA, deviceIndex = fromIntegral $ generatorDevice v'}
+                else
+                  Device {deviceType = CPU, deviceIndex = 0}
+          writeTVar generator $ Left (generatorSeed v', device)
+          return $ Right v'
+        Left v -> return (Left v)
     genPtr <- case mGenerator of
-      Just gen -> return gen
-      Nothing -> case device of
+      Right gen -> return gen
+      Left (seed, device) -> case device of
         Device CPU _ -> ATen.newCPUGenerator seed
         Device CUDA idx -> do
           gen <- ATen.newCUDAGenerator (fromIntegral idx)
           ATen.generator_set_current_seed gen seed
           return gen
     tensor <- cast3 func size genPtr options
-    nextSeed <- ATen.generator_current_seed genPtr
-    nextGenenerator <- newTVarIO (Just genPtr)
-    return (tensor, Generator device nextSeed nextGenenerator)
+    nextGenenerator <- newTVarIO (Right genPtr)
+    return (tensor, UnsafeGenerator nextGenenerator)
+  where
+    generatorIsCpu :: ForeignPtr ATen.Generator -> Bool
+    generatorIsCpu gen = unsafePerformIO $ cast1 ATen.generator_is_cpu gen
+    
+    generatorIsCuda :: ForeignPtr ATen.Generator -> Bool
+    generatorIsCuda gen = unsafePerformIO $ cast1 ATen.generator_is_cuda gen
+    
+    generatorDevice :: ForeignPtr ATen.Generator -> Int
+    generatorDevice gen = unsafePerformIO $ cast1 ATen.generator_get_device gen
+    
+    generatorSeed :: ForeignPtr ATen.Generator -> Word64
+    generatorSeed gen = fromIntegral (unsafePerformIO $ cast1 ATen.generator_get_device gen :: Int64)
```

Comment:
If these functions throw some exceptions, above `atomically` may be broken, but I think these functions  always succeed.

Reply:
We can not call IO function in `atomically`, so I use `unsafePerformIO`.